core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -304,6 +303,8 @@ class CachedDataFrame(df: Dataset[Row], var queryString: String,
       if (!hasLocalCallSite) {
         sc.clearCallSite()
       }
+      // clear the shuffle dependencies asynchronously after the execution.
+      clearCachedShuffleDeps(sc)
     }
```

Comment:
There are two other methods in Dataset that should be overridden for any required changes to collectWithHandler (had mentioned this in some other review but was ignored). See the calls to private method Dataset.collect(needCallback) which is also invoked by count() and head(). These two methods in Dataset should also be overridden in CachedDataFrame to use callbackWithHandler.

Can you take care of it in this PR because otherwise the shuffle cleanup will never happen for those two methods (e.g. if user does df.sql("...").count())?